### PR TITLE
Add cloud build Dockerfile and build steps

### PR DESCRIPTION
**Pull request by rpasricha**
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/fairing/92)

### CODE REVIEW

**Comment by rpasricha** [comment]
#83

File: Dockerfile.test [hunk 1]
```diff
@@ -0,0 +1,25 @@
+FROM python:3.5
+
+WORKDIR /opt/fairing
+
+# Install dependencies
+COPY setup.py requirements.txt /opt/fairing/
```
Comment (karthikv2k) [inline Dockerfile.test]: is this step redundant? line 23 would install deps too.

File: Dockerfile.test [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ENV PATH=/builder/google-cloud-sdk/bin/:$PATH
+
+# Install test dependencies
+RUN pip install pytest
+
+# Install Fairing
+COPY . /opt/fairing
+RUN python setup.py install
+
+CMD pytest
```
Comment (karthikv2k) [inline Dockerfile.test]: doesn't tests typically run with tests as working dir?

File: Dockerfile.test (hunk 1 above)
Comment (r2d4) [reply Dockerfile.test]: No, this copies over the requirement files first to download the dependencies for more optimal caching. 

If done together with line 23, then every time the source changes, we'd redownload all the dependencies

File: Dockerfile.test [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+WORKDIR /opt/fairing
+
+# Install dependencies
+COPY setup.py requirements.txt /opt/fairing/
+RUN pip install -e .
+
+# Install gcloud
+RUN mkdir -p /builder && \
+	wget -qO- https://dl.google.com/dl/cloudsdk/release/google-cloud-sdk.tar.gz | tar zxv -C /builder && \
+	CLOUDSDK_PYTHON="python2.7" /builder/google-cloud-sdk/install.sh --usage-reporting=false \
```
Comment (r2d4) [inline Dockerfile.test]: - python 3.5?
- pin to a specific gcloud version? 
- move this to the very top of the dockerfile, after WORKDIR

Dockerfile outline:
1. install gcloud
2. install pytest
3. copy setup.py and requirements, run pip install
4. copy over source
5. install fairing

File: Dockerfile.test (hunk 1 above)
Comment (rpasricha) [reply Dockerfile.test]: Yes, this helps us cache the dependencies in a different layer. The base image is not stored by default though, we'd need to add a separate step to pull the last image and use that as a cache - I'll go ahead and add that.

https://cloud.google.com/cloud-build/docs/speeding-up-builds

File: Dockerfile.test (hunk 2 above)
Comment (rpasricha) [reply Dockerfile.test]: pytest will extract tests from subdirectories, so runs successfully from the root project dir as well.

File: Dockerfile.test (hunk 3 above)
Comment (rpasricha) [reply Dockerfile.test]: Updated

File: Dockerfile.test (hunk 1 above)
Comment (rpasricha) [reply Dockerfile.test]: Updated to cache test image across builds

**Comment by rpasricha** [comment]
> For optimal caching:
> 
> Dockerfile outline:
> 
> 1. install gcloud
> 2. install pytest
> 3. copy setup.py and requirements, run pip install
> 4. copy over source
> 5. install fairing

Updated Dockerfile to have this order, thanks!

**Pull request merged**